activators/caila/src/main/kotlin/com/justai/jaicf/activator/caila/slotfilling/CailaSlotfillingHelper.kt

```diff
@@ -0,0 +1,148 @@
+package com.justai.jaicf.activator.caila.slotfilling
+
+
+import com.justai.jaicf.activator.caila.caila
+import com.justai.jaicf.activator.caila.client.CailaHttpClient
+import com.justai.jaicf.activator.caila.dto.CailaEntityMarkupData
+import com.justai.jaicf.activator.caila.dto.CailaKnownSlotData
+import com.justai.jaicf.api.BotRequest
+import com.justai.jaicf.context.ActionContext
+import com.justai.jaicf.context.ActivatorContext
+import com.justai.jaicf.context.BotContext
+import com.justai.jaicf.reactions.Reactions
+import com.justai.jaicf.slotfilling.*
+
+
+internal class CailaSlotfillingHelper(
+    private val cailaClient: CailaHttpClient,
+    private val cailaSlotFillingSettings: CailaSlotFillingSettings = CailaSlotFillingSettings.DEFAULT
+) {
+
+    fun process(
+        botContext: BotContext,
+        botRequest: BotRequest,
+        reactions: Reactions,
+        initialActivationContext: ActivatorContext?,
+        slotFiller: SlotFiller?
+    ): SlotFillingResult {
+        val ctx = restoreContext(botContext, initialActivationContext)
+        val required = ctx.requiredSlots
+        val known = ctx.knownSlots
+        if (known.map { it.name }.containsAll(required.map { it.name })) {
+            return SlotFillingSkipped
+        }
+
+        val actionContext = ActionContext(botContext, ctx.initialActivatorContext, botRequest, reactions, mutableListOf())
+        val filled = fillSlots(ctx, botRequest.input)
+        for (slot in required) {
+            if (slot.required && !known.any { k -> slot.name == k.name }) {
+                if (checkRetriesInterrupts(ctx, slot.name) || checkIntentInterrupts(botRequest.input)) {
+                    clearSlotfillingContext(botContext)
+                    return SlotFillingInterrupted
+                }
+                if (slot.name !in filled) {
+                    if (slotFiller != null && slotFiller.canHandle(slot.name)) {
+                        slotFiller.handle(slot.name, botContext, reactions)
+                    } else {
+                        with(actionContext) {
+                            reactions.sayRandom(*slot.prompts?.toTypedArray() ?: return SlotFillingSkipped)
+                        }
+                    }
+                    saveSlotFillingContext(botContext, ctx)
```

Comment:
Did you ever test this persistence with some context managers like mongo or mapdb?